metafacture-io/src/main/java/org/metafacture/io/HttpOpener.java

```diff
@@ -117,21 +203,109 @@ public void setHeader(final String key, final String value) {
         headers.put(key.toLowerCase(), value);
     }
 
+    /**
+     * Sets the HTTP request method.
+     *
+     * @param method the request method
+     */
+    public void setMethod(final Method method) {
+        this.method = method;
+    }
+
+    /**
+     * Sets the HTTP request URL.
+     *
+     * @param url the request URL
+     */
+    public void setUrl(final String url) {
+        this.url = url;
+    }
+
     @Override
-    public void process(final String urlStr) {
+    public void process(final String input) {
         try {
-            final URL url = new URL(urlStr);
-            final URLConnection con = url.openConnection();
-            headers.forEach(con::addRequestProperty);
-            String enc = con.getContentEncoding();
-            if (enc == null) {
-                enc = headers.get(ENCODING_HEADER);
+            final String requestUrl = getInput(input, url);
+            final String requestBody = getInput(input,
+                    body == null && method.getRequestHasBody() ? INPUT_DESIGNATOR : body);
+
+            final HttpURLConnection connection =
+                (HttpURLConnection) new URL(requestUrl).openConnection();
+
+            connection.setRequestMethod(method.name());
+            headers.forEach(connection::addRequestProperty);
+
+            if (requestBody != null) {
+                connection.setDoOutput(true);
+                connection.getOutputStream().write(requestBody.getBytes());
             }
-            getReceiver().process(new InputStreamReader(con.getInputStream(), enc));
+
+            final InputStream errorStream = connection.getErrorStream();
+            final InputStream inputStream = errorStream != null ?
+                getErrorStream(errorStream) : getInputStream(connection);
```

Comment:
There's no test coverage for `errorStream` yet. Wasn't able to come up with a way to trigger it.

Reply:
Maybe this could be achieved with `conn.setAllowUserInteraction(true)` (https://stackoverflow.com/questions/6070350/errorstream-in-httpurlconnection) but I am good with not testing this at all.

Reply:
This discussion seems to provide some useful info, thanks! I'll look into it.

(Basically: Read `errorStream` _after_ `inputStream` failed, not the other way around.)

Reply:
Fixed in b672731. Thanks :)